_posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd

```diff
@@ -0,0 +1,681 @@
+In the technical overview part, we build a simple unpooled model that describes every tank in isolation. Then, in the detailed walk-through, we'll see how to construct a _varying intercepts_ model that allows for information sharing between tanks.
+ 
+## Constructing models with `tfd_joint_distribution_sequential`
+
+`tfd_joint_distribution_sequential` represents a model as a list of conditional distributions.
+This is easiest to see on a real example, so we'll jump right in, creating an unpooled model of the tadpole data. 
+
+This is the how the model specification would look in _Stan_:
+
+```
+model{
+    vector[48] p;
+    a ~ normal( 0 , 1.5 );
+    for ( i in 1:48 ) {
+        p[i] = a[tank[i]];
+        p[i] = inv_logit(p[i]);
+    }
+    S ~ binomial( N , p );
+}
+```
+
+And here is `tfd_joint_distribution_sequential`:
+
+```{r}
+library(tensorflow)
+
+# make sure you have at least version 0.7 of TensorFlow Probability 
+# as of this writing, it is required of install the master branch:
+# install_tensorflow(version = "nightly")
+library(tfprobability)
+
+m1 <- tfd_joint_distribution_sequential(
+  list(
+    # normal prior of per-tank logits
+    tfd_multivariate_normal_diag(
+      loc = rep(0, n_tadpoles),
+      scale_identity_multiplier = 1.5),
+    # binomial distribution of survival counts
+    function(l)
+      tfd_independent(
+        tfd_binomial(total_count = n_start, logits = l),
+        reinterpreted_batch_ndims = 1
+      )
+  )
+)
+```
+
+The model consists of two distributions: Prior means and variances for the 48 tadpole tanks are specified by  `tfd_multivariate_normal_diag`; then `tfd_binomial` generates survival counts for each tank.
+Note how the first distribution is unconditional, while the second depends on the first. Note too how the second has to be wrapped in `tfd_independent` to avoid wrong broadcasting. (This is an aspect of `tfd_joint_distribution_sequential` usage that deserves to be documented more systematically, which is surely going to happen.^[As of today, lots of useful information is available in [Modeling with JointDistribution](https://github.com/tensorflow/probability/blob/master/tensorflow_probability/examples/jupyter_notebooks/Modeling_with_JointDistribution.ipynb) and [Multilevel Modeling Primer](https://github.com/tensorflow/probability/blob/master/tensorflow_probability/examples/jupyter_notebooks/Multilevel_Modeling_Primer.ipynb), but some experimentation may needed to adapt the -- numerous! -- examples to your needs.] Just think that this functionality was added to TFP `master` only three weeks ago!)
+
+As an aside, the model specification here ends up shorter than in _Stan_ as `tfd_binomial` optionally takes logits as parameters.
+
+As with every TFP distribution, a quick functionality check is sampling from the model:
```

Comment:
maybe "…you can do a quick functionality check by sampling from the model:"?

Reply:
yep